File: src/main/java/com/pedigree/editor/CommandStack.java
package com.pedigree.editor;

import java.util.ArrayDeque;
import java.util.Deque;

public class CommandStack {
    private final Deque<Command> undoStack = new ArrayDeque<>();
    private final Deque<Command> redoStack = new ArrayDeque<>();

    public void execute(Command command) {
        command.execute();
        undoStack.push(command);
        redoStack.clear();
    }

    public boolean canUndo() { return !undoStack.isEmpty(); }
    public boolean canRedo() { return !redoStack.isEmpty(); }

    public void undo() {
        if (!canUndo()) { return; }
        Command command = undoStack.pop();
        command.undo();
        redoStack.push(command);
    }

    public void redo() {
        if (!canRedo()) { return; }
        Command command = redoStack.pop();
        command.execute();
        undoStack.push(command);
    }

    public int getUndoSize() { return undoStack.size(); }
    public int getRedoSize() { return redoStack.size(); }
}

File: src/main/kotlin/com/family/tree/editor/CommandStack.kt
package com.family.tree.editor

import java.util.ArrayDeque

class CommandStack {
    private val undoStack: ArrayDeque<Command> = ArrayDeque()
    private val redoStack: ArrayDeque<Command> = ArrayDeque()

    fun execute(command: Command) {
        command.execute()
        undoStack.push(command)
        redoStack.clear()
    }

    fun canUndo(): Boolean = undoStack.isNotEmpty()
    fun canRedo(): Boolean = redoStack.isNotEmpty()

    fun undo() {
        if (!canUndo()) return
        val command = undoStack.pop()
        command.undo()
        redoStack.push(command)
    }

    fun redo() {
        if (!canRedo()) return
        val command = redoStack.pop()
        command.execute()
        undoStack.push(command)
    }

    fun getUndoSize(): Int = undoStack.size
    fun getRedoSize(): Int = redoStack.size
}
